crates/pathfinder/src/rpc.rs

```diff
@@ -73,339 +65,23 @@ Hint: If you are looking to run two instances of pathfinder, you must configure
-        use tracing::Instrument;
-
-        metrics::register_counter!("rpc_method_calls_total", "method" => method_name);
-
-        let method_callback = move |params: jsonrpsee::types::Params<'static>, context| {
-            // why info here? it's the same used in warp tracing filter for example.
-            let span = tracing::info_span!("rpc_method", name = method_name);
-            async move {
-                let input = params.parse::<Input>()?;
-                method(context, input).await.map_err(|err| {
-                    let rpc_err: error::RpcError = err.into();
-                    jsonrpsee::core::Error::from(rpc_err)
-                })
-            }
-            .instrument(span)
-        };
+        let context_v02 = (&self.api).into();
 
-        module
-            .register_async_method(method_name, method_callback)
-            .with_context(|| format!("Registering {method_name}"))?;
-
-        Ok(())
-    }
-
-    /// Registers a JSON-RPC method with the [RpcModule<RpcApi>](jsonrpsee::RpcModule).
-    ///
-    /// An example signature for `method` is:
-    /// ```ignore
-    /// async fn method(context: Arc<RpcApi>) -> Result<Ouput, Error>
-    /// ```
-    #[allow(dead_code)]
-    fn register_method_with_no_input<Output, Error, MethodFuture, Method>(
-        module: &mut jsonrpsee::RpcModule<RpcApi>,
-        method_name: &'static str,
-        method: Method,
-    ) -> anyhow::Result<()>
-    where
-        Output: 'static + ::serde::Serialize + Send + Sync,
-        Error: Into<error::RpcError>,
-        MethodFuture: std::future::Future<Output = Result<Output, Error>> + Send,
-        Method: (Fn(std::sync::Arc<RpcApi>) -> MethodFuture) + Copy + Send + Sync + 'static,
-    {
-        use anyhow::Context;
-        use tracing::Instrument;
-
-        metrics::register_counter!("rpc_method_calls_total", "method" => method_name);
-
-        let method_callback = move |_params, context| {
-            // why info here? it's the same used in warp tracing filter for example.
-            let span = tracing::info_span!("rpc_method", name = method_name);
-            async move {
-                method(context).await.map_err(|err| {
-                    let rpc_err: error::RpcError = err.into();
-                    jsonrpsee::core::Error::from(rpc_err)
-                })
-            }
-            .instrument(span)
-        };
+        let mut module_v01 = v01::RpcModuleWrapper::new(RpcModule::new(self.api));
+        v01::register_all_methods(&mut module_v01)?;
+        let module_v01: Methods = module_v01.into_inner().into();
 
-        module
-            .register_async_method(method_name, method_callback)
-            .with_context(|| format!("Registering {method_name}"))?;
+        let mut module_v02 = RpcModule::new(context_v02);
+        v02::register_all_methods(&mut module_v02)?;
+        let module_v02 = module_v02.into();
 
-        Ok(())
+        Ok(server
+            .start_with_paths([
```

Comment:
I did some attempts to expose a more natural api, where one would have to pass a closure that returns `impl jsonrpsee::(...)::Methods` and takes a `&hyper::Request` as argument, however I gave up as each instance of `Methods` has to run `initialize_resources` once at the very beginning, which has some consequences:
```rust=
// I gave up this api, because the user has to guarantee that mapper actually
// returns one of the methods it's given instead of ad-hoc creating some
// other instance
pub fn start_with<'a, MethodsVariants, MapperFn, IntoMethods, MethodsIter>(
	mut self,
        // methods is separated because each element undergoes `initialize_resources`
        // before ultimately being passes to `mapper` every time a request is made
	methods: MethodsVariants,
	mapper: MapperFn,
) -> Result<ServerHandle, Error>
where
	MethodsVariants: IntoIterator<Item = IntoMethods>,
	IntoMethods: Into<Methods>,
	MapperFn: Fn(
			&hyper::Request<hyper::Body>,
			std::sync::Arc<Vec<Methods>>,
		) -> std::ops::ControlFlow<hyper::Response<hyper::Body>, Methods>
		+ Send
		+ Sync
		+ 'static,
{
(...)
}
```